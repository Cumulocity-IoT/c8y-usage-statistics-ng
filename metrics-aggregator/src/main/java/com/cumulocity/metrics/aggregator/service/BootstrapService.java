package com.cumulocity.metrics.aggregator.service;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Service;
import com.cumulocity.microservice.subscription.model.MicroserviceSubscriptionAddedEvent;






/**
 * This is a service to fetch subtenants etc
 * 
 * @author Marco Stoffel
 *
 */

@Service
public class BootstrapService {

    @Autowired
    DeviceMetricsAggregationService deviceMetricsAggregationService;
 
    @Autowired
    StartupUtils utils;

    private static final Logger log = LoggerFactory.getLogger(BootstrapService.class);

    @EventListener
	public void initialize(MicroserviceSubscriptionAddedEvent event) {
		log.info("Tenant Sub: " + event.toString());
        utils.setCurrentTenant(event.getCredentials().getTenant());
        utils.getTenants();
        deviceMetricsAggregationService.createDailyDeviceStatistics();
	}

}
