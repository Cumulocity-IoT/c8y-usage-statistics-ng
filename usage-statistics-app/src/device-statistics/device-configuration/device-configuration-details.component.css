@media (max-width: 767px) {
  .alternate-grey {
    background-color: #F2F3F2;
  }
}
