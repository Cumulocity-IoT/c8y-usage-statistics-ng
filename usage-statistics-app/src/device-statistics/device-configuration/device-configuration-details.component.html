<div class="card card--fullpage">
  <c8y-title>
    {{ "Device Configuration" | translate }}
  </c8y-title>
  
  <div class="card-header hidden-xs separator">
    <div class="form-group m-0 col-sm-3">
      <label for="className">{{ "Classification" | translate}}</label>
    </div>
    <div class="form-group m-0 col-sm-3">
      <label for="className">{{ "Min daily MEA" | translate}}</label>
    </div>
    <div class="form-group m-0 col-sm-3">
      <label for="className">{{ "Max daily MEA" | translate}}</label>
    </div>
    <div class="form-group m-0 col-sm-3">
      <label for="className">{{ "Monthly device threshold" | translate}}</label>
    </div>
  </div>

  <div class="inner-scroll">
    <div class="card-block">
      <div [formGroup]="configurationForm">
        <ng-container formArrayName="configurations">
          <div class="row"
            *ngFor="
              let configurationListForm of configurations.controls;
              let first = first;
              let index = index;
              let last = last;
              let even = even;
            "
            [ngClass]="{'alternate-grey': even}"
          >
            <div [formGroupName]="index">
              <div class="form-group m-0 col-sm-3">
                <c8y-form-group
                  [hasError]="
                    configurationForm?.controls['configurations']?.controls[
                    index]?.controls?.className?.errors?.invalidClassname 
                  ">
                  <label class="visible-xs">{{ "Classification" | translate}}</label>
                  <input
                    formControlName="className"
                    type="string"
                    class="form-control"
                    (input)="duplicateClassnameValidate(
                      configurationForm?.controls['configurations']?.controls[index].controls?.className.value, index)"
                  />
                  <c8y-messages>
                    <c8y-message
                      *ngIf="
                        configurationForm?.controls['configurations']?.controls[
                        index]?.controls?.className?.errors?.invalidClassname
                      "
                    >
                      {{ "Classification names cannot be same." | translate}}
                    </c8y-message>
                  </c8y-messages>
                </c8y-form-group>
              </div>

              <!--First row's minMEA value should be always 0-->
              <div class="form-group m-0 col-sm-3" *ngIf="first">
                <c8y-form-group>
                  <label class="visible-xs">{{ "Min daily MEA" | translate}}</label>
                  <input
                    formControlName="avgMinMea"
                    type="number"
                    class="form-control"
                    placeholder="0"
                    disabled
                  />
                </c8y-form-group>
              </div>

              <!--When it is not First row-->
              <div class="form-group m-0 col-sm-3" *ngIf="!first">
                <c8y-form-group>
                  <label class="visible-xs">{{ "Min daily MEA" | translate}}</label>
                  <input
                    formControlName="avgMinMea"
                    type="number"
                    class="form-control"
                    required
                    disabled
                  />
                </c8y-form-group>
              </div>

              <!--If not last row, then maxMEA will be default/saved value-->
              <div class="form-group m-0 col-sm-3" *ngIf="!last">
                <c8y-form-group
                  [hasError]="
                    avgMaxMea !== '' &&
                    (configurationForm?.controls['configurations']?.controls[
                      index
                    ].controls?.avgMaxMea?.errors?.greaterMin ||
                      configurationForm?.controls['configurations']?.controls[
                        index
                      ].controls?.avgMaxMea?.errors) ||
                  configurationForm?.controls['configurations']?.controls[
                    index
                  ].controls?.avgMaxMea?.errors?.greaterMax ||
                  configurationForm?.controls['configurations']?.controls[
                    index
                  ].controls?.avgMaxMea?.errors?.required
                  "
                >
                  <label class="visible-xs">{{ "Max daily MEA" | translate}}</label>
                  <input
                    formControlName="avgMaxMea"
                    type="number"
                    class="form-control"
                    min="0"
                    required
                    (input)="isValidValue()"
                  />
                  <c8y-messages>
                    <c8y-message
                      *ngIf="
                        avgMaxMea !== '' &&
                        configurationForm?.controls['configurations']?.controls[
                          index
                        ].controls?.avgMaxMea?.errors?.greaterMin
                      "
                    >
                      {{ "Max value should be greater than Min value." | translate}}
                    </c8y-message>
                    <c8y-message
                      *ngIf="
                        avgMaxMea !== '' &&
                        configurationForm?.controls['configurations']?.controls[
                          index
                        ].controls?.avgMaxMea?.errors?.min ||
                        configurationForm?.controls['configurations']?.controls[
                        index
                      ].controls?.avgMaxMea?.errors?.required
                    "
                  >
                    {{ "Max value should be a positive number." | translate }}
                  </c8y-message>
                  <c8y-message
                    *ngIf="
                        avgMaxMea !== '' &&
                        configurationForm?.controls['configurations']?.controls[
                          index
                        ].controls?.avgMaxMea?.errors?.greaterMax
                  "
                  translate
                    >
          Max value {{configurationForm?.controls['configurations']?.controls[
                      index
                    ].controls?.className?.value}} should be less than max value of {{configurationForm?.controls['configurations']?.controls[
                    index+1
                  ].controls?.className?.value}}.
                    </c8y-message>
                  </c8y-messages>
                </c8y-form-group>
              </div>

              <!--Last row's maxMEA value should be INFINITY always-->
              <div class="form-group m-0 col-sm-3" *ngIf="last">
                <c8y-form-group>
                  <label class="visible-xs">{{ "Max daily MEA" | translate}}</label>
                  <input
                    formControlName="avgMaxMea"
                    type="text"
                    class="form-control"
                    placeholder="INFINITY"
                    disabled
                  />
                </c8y-form-group>
              </div>

              <div class="form-group m-0 col-sm-3">
                <c8y-form-group
                  [hasError]="
                    monthlyThreshold !== '' &&
                    configurationForm?.controls['configurations']?.controls[
                      index
                    ].errors?.invalidThreshold
                  "
                >
                  <label class="visible-xs">{{ "Monthly device threshold" | translate}}</label>
                  <input
                    formControlName="monthlyThreshold"
                    type="number"
                    class="form-control"
                    min="0"
                  />
                  <c8y-messages>
                    <c8y-message
                      *ngIf="
                        monthlyThreshold !== '' &&
                        configurationForm?.controls['configurations']?.controls[
                          index
                        ].errors?.invalidThreshold
                      "
                    >
                      {{ "Threshold value should be a positive number." | translate}}
                    </c8y-message>
                  </c8y-messages>
                </c8y-form-group>
              </div>
            </div>
          </div>
        </ng-container>
      </div>
    </div>
  </div>

  <div class="card-footer separator">
    <button
      type="button"
      class="btn btn-default"
      title="{{ 'Reset' | translate }}"
      (click)="resetToDefault()"
    >
      <span>{{ "Reset" | translate }}</span>
    </button>
    <button
      (click)="onCancel()"
      type="button"
      class="btn btn-default"
      title="{{ 'Cancel' | translate }}"
    >
      <span>{{ "Cancel" | translate }}</span>
    </button>
    <button
      type="button"
      (click)="onSubmit()"
      class="btn btn-primary"
      title="{{ 'Save' | translate }}"
      [disabled]="!isSaveEnabled"
    >
      <span>{{ "Save" | translate }}</span>
    </button>
  </div>
</div>
