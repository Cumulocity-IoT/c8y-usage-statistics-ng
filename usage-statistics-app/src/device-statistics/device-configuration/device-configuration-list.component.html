<div class="col-xs-12 col-sm-12 col-md-8">
  <div class="card--fullpage">
    <c8y-title>
      {{ "Device Configuration" | translate }}
    </c8y-title>
  
    <c8y-action-bar-item [placement]="'right'" translate>
      <div class="p-t-8 p-b-8">
        <button
          type="button"
          class="btn btn-link"
          title="{{ 'Edit' | translate }}"
          (click)="onEdit()"
        >
          <i c8yicon="pencil" class="m-r-4 dlt-c8y-icon-pencil"></i>
          {{ "Edit" | translate }}
        </button>
      </div>
    </c8y-action-bar-item>
  
    <div class="content-fullpage d-flex d-col shadow1">
      <table class="table">
        <thead class="page-sticky-header">
          <tr>
            <th
              class="col-4 p-l-16"
              title="{{ 'Classification Name' | translate }}"
            >
              {{ "Classification Name" | translate }}
            </th>
            <th
              class="col-4"
              title="{{ 'Daily MEA range' | translate }}"
            >
              {{ "Daily MEA range" | translate }}
              <button
                type="button"
                class="btn btn-clean text-primary"
                triggers="focus"
                placement="auto"
                popover="{{ 'Mean daily average (MEA)' | translate }}"
              >
                <i
                  c8y-icon="question-circle-o"
                  class="dlt-c8y-icon-question-circle-o"
                ></i>
              </button>
            </th>
            <th
              class="col-4"
              title="{{ 'Monthly Device Threshold' | translate }}"
            >
              {{ "Monthly Device Threshold" | translate }}
            </th>
          </tr>
        </thead>
        <tbody *ngFor="let data of data; 
                      let first = first;
                      let last = last">
          <tr>
            <td class="p-l-16">{{ data.className }}</td>
            <td *ngIf="!last"> &ge; {{ data.avgMinMea }} {{ " and " | translate }} &lt; {{ data.avgMaxMea }} {{ " per day " | translate }}</td>
            <td *ngIf="last"> &ge; {{ data.avgMinMea }} {{ " per day " | translate }}</td>
            <td>{{ data.monthlyThreshold }}</td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>