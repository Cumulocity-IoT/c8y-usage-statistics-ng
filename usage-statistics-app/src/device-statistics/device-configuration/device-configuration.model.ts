export interface ConfigOptions {
    category: string;
    value: string;
    key: string;
  }

export interface Configuration {
  className: string;
  avgMinMea: number;
  avgMaxMea: number | string;
  monthlyThreshold: number;
}
