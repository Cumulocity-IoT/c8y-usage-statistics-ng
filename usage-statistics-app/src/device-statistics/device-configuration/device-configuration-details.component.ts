import { Component, OnInit } from "@angular/core";
import {
  AbstractControl,
  FormArray,
  FormBuilder,
  FormControl,
  FormGroup,
  ValidationErrors,
  ValidatorFn,
  Validators,
} from "@angular/forms";
import { Router } from "@angular/router";
import { AlertService, GainsightService, gettext } from "@c8y/ngx-components";
import { Configuration } from "./device-configuration.model";
import { DeviceConfigurationService } from "./device-configuration.service";

@Component({
  selector: "app-device-configuration-details",
  templateUrl: "./device-configuration-details.component.html",
  styleUrls: ['./device-configuration-details.component.css']
})
export class DeviceConfigurationComponentDetails implements OnInit {
  configData: Configuration[];
  errors = [];
  isLoading = true;
  isSaveEnabled = false;
  configControls: AbstractControl;

  configurationListForm = this.fb.group({
    className: new FormControl("", [Validators.required]),
    avgMinMea: new FormControl("", [Validators.required, Validators.min(0)]),
    avgMaxMea: new FormControl("", [Validators.min(0)]),
    monthlyThreshold: new FormControl(""),
  });
 
  configurationForm: any = this.fb.group({
    configurations: this.fb.array([]),
  });

  constructor(
    private deviceConfigService: DeviceConfigurationService,
    private fb: FormBuilder,
    private router: Router,
    private alert: AlertService,
    private gainsightService: GainsightService
  ) {}

  ngOnInit(): void {
    this.getConfigurationOptions();
  }

  /**
   * gets the saved configuration values from deviceStatisticsConfigurationStore
   * if not present in the store, then fetch it using the service
   */
  async getConfigurationOptions() {
    this.configData =
      this.deviceConfigService.deviceStatisticsConfigurationStore ||
      (await this.deviceConfigService.getConfigurationOptions());

    this.configurationForm = this.fb.group({
      configurations: this.fb.array(
        this.configData.map((config) => this.createConfiguration(config))
      ),
    });
  }

  /**
   * determines if current avg max mea value is a valid value
   * below are some invalid scenarios:
   * current avg max mea value < 0 -> invalid
   * current avg max mea value >= next row's avg max mea value -> invalid
   * current avg min mea value >= current avg max mea value -> invalid
   */
   isValidValue() {
    let nextMax = '';
    for (let j = 0; j < this.configControls.value.length-1; j++) {
      const valueControlMin = this.configControls["controls"][j].value.avgMinMea;
      const valueControlMax = this.configControls["controls"][j].value.avgMaxMea;
      const currentMaxValue = this.configControls?.['controls'][j].get('avgMaxMea').value;
      this.configurationForm?.controls?.configurations?.['controls'][j+1].get('avgMinMea').setValue(currentMaxValue);
      nextMax = this.configControls["controls"][j+1].value.avgMaxMea;
      if ( valueControlMax < 0) {
        this.configControls?.['controls'][j].get('avgMaxMea').setErrors({ 'customMin': true });
        this.errors.push({ customMin: true });
      } else if (currentMaxValue >= nextMax) {
        this.configControls?.['controls'][j].get('avgMaxMea').setErrors({'greaterMax': true});
        this.errors.push({ greaterMax: true });
      } else if (valueControlMin >= valueControlMax) {
        this.configControls?.['controls'][j].get('avgMaxMea').setErrors({'greaterMin': true});
        this.errors.push({ greaterMin: true });
      } else {
        this.configControls?.['controls'][j].get('avgMaxMea').setErrors(null);
      }
    }
  }

  duplicateClassnameValidate(className : string, index: number) {
    for (let j = 0; j < this.configControls.value.length-1; j++) {
      const currentClassname = this.configControls?.['controls'][j].get('className').value;
      if (className === currentClassname && index !== j) {
        this.configControls?.['controls'][index].get('className').setErrors({'invalidClassname': true});
        return;
      } else {
        this.configControls?.['controls'][index].get('className').setErrors(null);
        this.configControls?.['controls'][j].get('className').setErrors(null);
      }
    }
  }

  /**
   * creates formgroup with the obtained result from the store variable/get service call
   * @param config row of configuration values obtained from store variable/get service call
   * @returns formgroup
   */
  createConfiguration(config): FormGroup {
    const formgroupConfig = this.fb.group(
      {
        className: [config.className],
        avgMinMea: [config.avgMinMea],
        avgMaxMea: [config.avgMaxMea],
        monthlyThreshold: [config.monthlyThreshold],
      },
      { validators: this.formArrayValidator() }
    );

    return formgroupConfig;
  }

  /**
   * checks if threshold value is less than 0
   * also disables save button when error occurs
   * @returns error in key value pair or null
   */
  formArrayValidator(): ValidatorFn {
    const a = this.configurationForm.controls["configurations"];
    return (a: AbstractControl): ValidationErrors | null => {
      this.configControls = this.configurationForm?.controls["configurations"];

      for (let i = 0; i < this.configControls.value.length; i++) {
        const monthlyThreshold =
          this.configControls["controls"][i].value.monthlyThreshold;
        if (monthlyThreshold && monthlyThreshold < 0) {
          this.isSaveEnabled = false; // disable save button
          return { invalidThreshold: true };
        } else {
          this.isSaveButtonEnabled();
        }
      }
      // when "errors" array is not empty, disable save button
      if (this.errors.length !== 0 || !this.configurationForm.dirty) {
        this.isSaveEnabled = false;
      } else {
        this.isSaveEnabled = true;
      }
      return null;
    };
  }

  /**
   * this function determines if save button is enabled or not
   * if user enters invalid value, "errors" array will not be empty
   */
  isSaveButtonEnabled() {
    const configurationControls = this.configControls["controls"];
    this.errors = [];
    for (let i = 0; i < configurationControls.length; i++) {
      if (
        configurationControls &&
        configurationControls[i] &&
        (configurationControls[i]["controls"]?.avgMinMea.errors ||
          configurationControls[i]["controls"]?.avgMaxMea.errors)
      ) {
        this.errors.push({ invalid: true });
      }
    }
  }

  /**
   * getter function to get configuration details
   */
  get configurations() {
    return this.configurationForm.controls["configurations"] as FormArray;
  }

  /**
   * Resets the values to default ones and
   * redirects to the main list page
   */
  resetToDefault() {
    this.gainsightService.triggerEvent(
      "usageStatistics",
      { action: "reset", name: "DeviceConfigurationReset" }
    );
    const defaultConfigValue = this.deviceConfigService.defaultConfig;
    this.deviceConfigService
      .setOptions(defaultConfigValue)
      .then(() => {
        this.alert.success(gettext('Device class configuration values are reset to default values.'));
        this.onCancel()
      });
  }

  /**
   * Sorts the records according to avg min mea value.
   * In future when we provide functionality for adding
   * new rows,this will help in getting the rows in
   * sequential manner after saving.
   */
  onSubmit() {
    this.gainsightService.triggerEvent(
      "usageStatistics",
      { action: "save", name: "DeviceConfigurationSave" }
    );
    const value = this.configurations.value;
    const sortedValues = value.sort((a, b) => a.avgMinMea - b.avgMinMea);
    this.save(sortedValues);
  }

  /**
   * saves the configuration values to the service
   * @param options sorted configuration values
   * @returns status code
   */
  async save(options: Configuration[]) {
    try {
      const res = await this.deviceConfigService.setOptions(options);
      if (res && res.status !== 200) {
        const data = res.json ? await res.json() : undefined;
        this.alert.danger(gettext('Could not update device class configuration'), data);
        return Promise.reject();
      } else {
        this.alert.success(gettext('Device class configuration values are updated successfully.'));
        this.onCancel();
        return Promise.resolve();
      }
    } catch (ex) {
      this.alert.danger(gettext('Could not update device class configuration'), ex);
      return Promise.reject();
    }
  }

  /**
   * onCancel, go back to the list of configurations saved
   */
  onCancel() {
    this.router.navigateByUrl(`device-statistics/device-configuration-list`);
  }
}
