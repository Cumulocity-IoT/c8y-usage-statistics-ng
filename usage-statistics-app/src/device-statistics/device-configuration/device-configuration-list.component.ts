import { Component } from "@angular/core";
import { Router } from "@angular/router";
import { IFetchResponse } from "@c8y/client";
import { AlertService, gettext } from "@c8y/ngx-components";
import { ConfigOptions, Configuration } from "./device-configuration.model";
import { DeviceConfigurationService } from "./device-configuration.service";

@Component({
  selector: "app-device-configuration-list",
  templateUrl: "./device-configuration-list.component.html",
})
export class DeviceConfigurationComponentList {
  data: Configuration[];
  configData: ConfigOptions;
  isLoading = false;

  constructor(
    private deviceConfigService: DeviceConfigurationService,
    private alert: AlertService,
    private router: Router
  ) {
  }

  ngOnInit() {
    this.getConfigurationOptions();
  }

  /**
   * gets the saved device configuration options.
   * if no saved configuration present, then default configuration is loaded
   */
  async getConfigurationOptions() {
    this.data = await this.deviceConfigService.getConfigurationOptions();    
  }

  /**
   * this function helps to modify the existing device configuration option
   * @param options
   * @returns promise
   */
  async updateConfigurationOptions(options: Configuration[]) {
    try {
      const res = await this.deviceConfigService.setOptions(options);
      if (res && res.status !== 200) {
        const data = res.json ? await res.json() : undefined;
        this.alert.addServerFailure({ data, res });

        return Promise.reject(gettext('Setting device configuration option failed!'));
      } else {
        return Promise.resolve();
      }
    } catch (ex) {
      this.alert.addServerFailure(ex);

      return Promise.reject();
    }
  }

  /**
   * this function resets the configuration to default values
   */
  setDefaultConfiguration() {
    const defaultConfigValue = this.deviceConfigService.defaultConfig;
    this.updateConfigurationOptions(defaultConfigValue);
    this.getConfigurationOptions();
  }

  /**
   * this redirects to configuration details page
   */
  onEdit() {
    this.router.navigateByUrl(`device-statistics/device-configuration-details`);
  }
}
