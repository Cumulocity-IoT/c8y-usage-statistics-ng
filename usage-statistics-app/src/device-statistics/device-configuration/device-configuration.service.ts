import { Injectable } from "@angular/core";
import { FetchClient, IFetchOptions, IFetchResponse, IResult } from "@c8y/client";
import { Configuration } from "./device-configuration.model";

@Injectable({
  providedIn: "root",
})
export class DeviceConfigurationService {
  private readonly microserviceUrl: string = "/tenant/options";
  private readonly header: any = { "Content-Type": "application/json" };

  defaultConfig: Configuration[];
  deviceStatisticsConfigurationStore: Configuration[];
  previousConfiguration: Configuration[];

  constructor(private client: FetchClient) {
    this.defaultConfig = [
      {
        className: "Class A",
        avgMinMea: 0,
        avgMaxMea: 24,
        monthlyThreshold: null,
      },
      {
        className: "Class B",
        avgMinMea: 24,
        avgMaxMea: 144,
        monthlyThreshold: null,
      },
      {
        className: "Class C",
        avgMinMea: 144,
        avgMaxMea: 1440,
        monthlyThreshold: null,
      },
      {
        className: "Class D",
        avgMinMea: 1440,
        avgMaxMea: 8640,
        monthlyThreshold: null,
      },
      {
        className: "Class E",
        avgMinMea: 8640,
        avgMaxMea: 86400,
        monthlyThreshold: null,
      },
      {
        className: "Class F",
        avgMinMea: 86400,
        avgMaxMea: "INFINITY",
        monthlyThreshold: null,
      },
    ];
    this.deviceStatisticsConfigurationStore = this.defaultConfig;
  }

  async getConfigurationOptions() {
    //const response = await this.getOptions();
    // this.previousConfiguration = this.deviceStatisticsConfigurationStore;
    // this.deviceStatisticsConfigurationStore = JSON.parse(response.data.value);
    return this.deviceStatisticsConfigurationStore;
  }

  async getOptions(): Promise<IResult<any>> {
    const options: IFetchOptions = {
      method: "GET",
      headers: this.header,
    };
    const option = {
      category: "configuration",
      key: "device.statistics.class.details",
    };

    const url = `${this.microserviceUrl}/${option.category}/${option.key}`;

    const res = await this.client.fetch(url, options);
    let data = await res.json();
    if (res.status !== 200) {
      await this.setOptions(this.defaultConfig);
      data.value = JSON.stringify(this.defaultConfig);
     }

    return { res, data };
  }

  async setOptions(configValues: Configuration[]): Promise<IFetchResponse> {
    const configData = {
      category: "configuration",
      key: "device.statistics.class.details",
      value: JSON.stringify(configValues),
    };
    const options: IFetchOptions = {
      method: "POST",
      headers: this.header,
      body: JSON.stringify(configData),
    };
    return await this.client.fetch(`${this.microserviceUrl}`, options);
  }
}
