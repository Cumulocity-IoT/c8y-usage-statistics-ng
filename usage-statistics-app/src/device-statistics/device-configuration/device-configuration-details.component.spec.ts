import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, fakeAsync, TestBed } from '@angular/core/testing';
import { FormBuilder, FormGroup } from '@angular/forms';
import { Router } from '@angular/router';
import { RouterTestingModule } from '@angular/router/testing';
import { AlertService, CoreModule, GainsightService } from '@c8y/ngx-components';
import { DeviceConfigurationComponentDetails } from './device-configuration-details.component';
import { DeviceConfigurationService } from './device-configuration.service';

describe('Test DeviceConfigurationComponentDetails', () => {
  let component: DeviceConfigurationComponentDetails;
  let fixture: ComponentFixture<DeviceConfigurationComponentDetails>;
  let deviceConfigurationService: DeviceConfigurationService;
  let alertService: AlertService;
  let router: Router;
  let fb: FormBuilder;
  let gainsightService: GainsightService;
  let defaultConfig;
  let configurationForm;
  let configControls;

  const deviceConfigurationServiceMock = {
    getConfigurationOptions: jest.fn(),
    setOptions: jest.fn(),
  }

  const alertServiceMock = {
    addServerFailure: jest.fn()
  }

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [RouterTestingModule, CoreModule.forRoot()],
      declarations: [DeviceConfigurationComponentDetails],
      schemas: [NO_ERRORS_SCHEMA],
      providers: [
        { provide: DeviceConfigurationService, useValue: deviceConfigurationServiceMock },
        { provide: AlertService, useValue: alertServiceMock },
    ]
    });
    fixture = TestBed.createComponent(DeviceConfigurationComponentDetails);
    component = fixture.componentInstance;
    deviceConfigurationService = TestBed.inject(DeviceConfigurationService);
    alertService = TestBed.inject(AlertService);
    router = TestBed.inject(Router);
    fb = TestBed.inject(FormBuilder),
    gainsightService = TestBed.inject(GainsightService)
  });

  beforeEach(() => {
    defaultConfig = [
      {
        className: "Class A",
        avgMinMea: 0,
        avgMaxMea: 24,
        monthlyThreshold: null,
      },
      {
        className: "Class B",
        avgMinMea: 24,
        avgMaxMea: 144,
        monthlyThreshold: null,
      },
      {
        className: "Class C",
        avgMinMea: 144,
        avgMaxMea: 1440,
        monthlyThreshold: null,
      },
      {
        className: "Class D",
        avgMinMea: 1440,
        avgMaxMea: 8640,
        monthlyThreshold: null,
      },
      {
        className: "Class E",
        avgMinMea: 8640,
        avgMaxMea: 86400,
        monthlyThreshold: null,
      },
      {
        className: "Class F",
        avgMinMea: 86400,
        avgMaxMea: "INFINITY",
        monthlyThreshold: null,
      },
    ];
    configurationForm = fb.group({
      configurations: fb.array(
        defaultConfig.map((config) => createConfiguration(config))
      ),
    });
  })

  test('should be defined', () => {
    expect(fixture).toBeDefined();
  });

  test('should load expected component', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('should get configuration option' , () => {
    
    test('should call get configuration option funtion on init', () => {
      fixture.detectChanges();
      
      jest.spyOn(component, 'getConfigurationOptions');
      jest.spyOn(deviceConfigurationService, 'getConfigurationOptions').mockResolvedValue(defaultConfig);

      component.ngOnInit();

      expect(component.getConfigurationOptions).toHaveBeenCalled();
    })

    test('should call getConfigurationOptions from service page if not present in deviceStatisticsConfigurationStore', () => {
      jest.spyOn(deviceConfigurationService, 'getConfigurationOptions');
      component.getConfigurationOptions();
      expect(deviceConfigurationService.getConfigurationOptions).toHaveBeenCalled();

    })
  })

  describe('should verify the validations', () => {
    beforeEach(() => {
      configControls = configurationForm.controls["configurations"];
      component.configurationForm = configurationForm;
      component.configControls = configControls;
    })

    test('should give customMin error when max mea value is negative', () => {
      component.configControls["controls"][0].value.avgMaxMea = -5 ;
      component.configurationForm.updateValueAndValidity();
      component.isValidValue();
      let avgMaxMeaError = configControls['controls'][0].get('avgMaxMea').errors;
      expect(avgMaxMeaError).toEqual({'customMin': true});
    })

    test('should give greaterMax error when max mea of one row is greater than max mea of next row', () => {
      /*setting first row max value to 500 will make it greater than second row's default max value 144, there by giving error*/
      component.configControls['controls'][0].get('avgMaxMea').value = 500;
      component.configurationForm.updateValueAndValidity();
      component.isValidValue();
      let avgMaxMeaError = configControls['controls'][0].get('avgMaxMea').errors;
      expect(avgMaxMeaError).toEqual({'greaterMax': true});
    })

    test('should give greaterMin error when min mea is greater than max mea', () => {
      component.configControls["controls"][0].value.avgMinMea = 100;
      component.configControls["controls"][0].value.avgMaxMea = 10;
      component.configurationForm.updateValueAndValidity();
      component.isValidValue();
      let avgMaxMeaError = configControls['controls'][0].get('avgMaxMea').errors;
      expect(avgMaxMeaError).toEqual({'greaterMin': true});
    })

    test('should disable save button if validation of the form fails', () => {
      component.configControls["controls"][0].value.avgMaxMea = -5 ;
      component.configurationForm.updateValueAndValidity();
      component.isValidValue();
      component.formArrayValidator();

      expect(component.errors.length).toEqual(1);
      expect(component.isSaveEnabled).toEqual(false);
    })

    test('save button should be disabled on load', () => {
      component.configurationForm.updateValueAndValidity();
      component.isValidValue();
      component.formArrayValidator();
      expect(component.errors.length).toEqual(0);
      expect(component.isSaveEnabled).toEqual(false);
    })
  })

  function createConfiguration(config): FormGroup {
    const formgroupConfig = fb.group(
      {
        className: [config.className],
        avgMinMea: [config.avgMinMea],
        avgMaxMea: [config.avgMaxMea],
        monthlyThreshold: [config.monthlyThreshold],
      }
    );
  
    return formgroupConfig;
  }
})

