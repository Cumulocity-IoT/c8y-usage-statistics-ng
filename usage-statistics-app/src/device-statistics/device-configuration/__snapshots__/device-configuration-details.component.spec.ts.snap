// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test DeviceConfigurationComponentDetails should load expected component 1`] = `
<app-device-configuration-details
  alert={[Function Object]}
  configurationForm={[Function FormGroup]}
  configurationListForm={[Function FormGroup]}
  deviceConfigService={[Function Object]}
  errors={[Function Array]}
  fb={[Function FormBuilder]}
  gainsightService={[Function GainsightService]}
  isLoading={[Function Boolean]}
  isSaveEnabled="false"
  router={[Function Router]}
>
  <div
    class="card card--fullpage"
  >
    <c8y-title>
    </c8y-title>
    <div
      class="card-header hidden-xs separator"
    >
      <div
        class="form-group m-0 col-sm-3"
      >
        <label
          for="className"
        >
        </label>
      </div>
      <div
        class="form-group m-0 col-sm-3"
      >
        <label
          for="className"
        >
        </label>
      </div>
      <div
        class="form-group m-0 col-sm-3"
      >
        <label
          for="className"
        >
        </label>
      </div>
      <div
        class="form-group m-0 col-sm-3"
      >
        <label
          for="className"
        >
        </label>
      </div>
    </div>
    <div
      class="inner-scroll"
    >
      <div
        class="card-block"
      >
        <div>
        </div>
      </div>
    </div>
    <div
      class="card-footer separator"
    >
      <button
        class="btn btn-default"
        type="button"
      >
        <span>
        </span>
      </button>
      <button
        class="btn btn-default"
        type="button"
      >
        <span>
        </span>
      </button>
      <button
        class="btn btn-primary"
        type="button"
      >
        <span>
        </span>
      </button>
    </div>
  </div>
</app-device-configuration-details>
`;
